### DISCUSSION TITLE
Consumer identification from OpenId/Keycloak access token

### DISCUSSION THREAD

**Discussion by singhjoga** [apache/apisix#5547] [Q&A] [answered]
Hi Guys,
I have an existing API which uses Keycloak to authenticate requests using Bearer tokens (generated using Client Credentials flow). I want to use APISIX to implement Rate Limits for certain consumers. Is there a way to identify the consumer (client id used for token generation) from the access token generated by Keycloak?

Thanks in advance.
Best Regards,

Joga Singh

**Comment by spacewander** [comment, marked as answer]
The consumer needs to be configured in APISIX. Since the access token is generated by Keycloak, how can we match it with the consumer?

**Reply by singhjoga** [reply]
@spacewander thanks for your reply. I thought maybe there is some way using Keycloak plugin or something.  From your answer now I know that it is not possible.

**Reply by juzhiyuan** [reply]
After searching through docs and codes, Apache APISIX cannot verify consumers by accesstoken from outside Keycloack for now IMO, why not have a try to ask in the mailing list? Here is the subscription guide: https://apisix.apache.org/docs/general/subscribe-guide

Maybe there have some community users who have the same requirements and find your answer 😄

**Comment by juzhiyuan** [comment]
> I want to use APISIX to implement Rate Limits for certain consumers.

Hi @singhjoga, after reading the concepts here[1], here is my understanding:

1. Create Consumers with different usernames;
2. Create a Route with Traffic Limit plugin and also **Consumer Restriction plugin[2]**

I'm not sure if this meets your needs? and also cc @bzp2010 to have a confirmation 😄

[1] https://apisix.apache.org/docs/apisix/architecture-design/consumer
[2] https://apisix.apache.org/docs/apisix/plugins/consumer-restriction/

**Reply by singhjoga** [reply]
@juzhiyuan thanks for your answer. I knew this already. I was thinking to find something using the access tokens itself.

**Comment by starsz** [comment]
Hi @singhjoga 

> I want to use APISIX to implement Rate Limits for certain consumers

It's a little difficult now, but `implement Rate Limits for each consumers` is not difficult.
You can try to write some variables (like name, sub id) into ctx.var at here: 
https://github.com/apache/apisix/blob/8f0b066c86257ad6af19f9b3b7e209ece95d17c9/apisix/plugins/openid-connect.lua#L301-L320
And then we can add a limit-* plugin with the config of those variables.

As for ` implement Rate Limits for certain consumers`

I create issue here: https://github.com/apache/apisix/issues/5634
If we can support get the key from function in limit-* plugins, then we can add function to get the key of consumer and limit for certain consumers.
